include/deal.II/matrix_free/portable_evaluation_kernels.h

```diff
@@ -0,0 +1,620 @@
+          Kokkos::subview(shared_data->scratch_pad,
+                          Kokkos::make_pair(0, scratch_size));
+
+        EvaluatorTensorProduct<EvaluatorVariant::evaluate_general,
+                               dim,
+                               fe_degree + 1,
+                               n_q_points_1d,
+                               Number>
+          eval(shared_data->team_member,
+               data->shape_values,
+               data->shape_gradients,
+               data->co_shape_gradients,
+               scratch_for_eval);
+
+        for (unsigned int c = 0; c < n_components; ++c)
+          {
+            auto u      = Kokkos::subview(shared_data->values, Kokkos::ALL, c);
+            auto grad_u = Kokkos::subview(shared_data->gradients,
+                                          Kokkos::ALL,
+                                          Kokkos::ALL,
+                                          c);
+
+            // apply derivatives in collocation space
+            if (integration_flag & EvaluationFlags::gradients)
+              {
+                if constexpr (dim == 1)
+                  {
+                    if (integration_flag & EvaluationFlags::values)
+                      eval.template co_gradients<0, false, true, false>(
+                        Kokkos::subview(grad_u, Kokkos::ALL, 0), u);
+                    else
+                      eval.template co_gradients<2, false, false, false>(
+                        Kokkos::subview(grad_u, Kokkos::ALL, 2), u);
+                  }
+                else if constexpr (dim == 2)
+                  {
+                    if (integration_flag & EvaluationFlags::values)
+                      eval.template co_gradients<1, false, true, false>(
+                        Kokkos::subview(grad_u, Kokkos::ALL, 1), u);
+                    else
+                      eval.template co_gradients<1, false, false, false>(
+                        Kokkos::subview(grad_u, Kokkos::ALL, 1), u);
+                    eval.template co_gradients<0, false, true, false>(
+                      Kokkos::subview(grad_u, Kokkos::ALL, 0), u);
+                  }
+                else if constexpr (dim == 3)
+                  {
+                    if (integration_flag & EvaluationFlags::values)
+                      eval.template co_gradients<2, false, true, false>(
+                        Kokkos::subview(grad_u, Kokkos::ALL, 2), u);
+                    else
+                      eval.template co_gradients<2, false, false, false>(
+                        Kokkos::subview(grad_u, Kokkos::ALL, 2), u);
+                    eval.template co_gradients<1, false, true, false>(
+                      Kokkos::subview(grad_u, Kokkos::ALL, 1), u);
+                    eval.template co_gradients<0, false, true, false>(
+                      Kokkos::subview(grad_u, Kokkos::ALL, 0), u);
+                  }
+                else
+                  Kokkos::abort("dim must not exceed 3!");
```

Comment:
and here

Reply:
Depending on the implementation of `Kokkos::abort`, a `static_assert` should save some registers.